Outreachy: getting familiar with the development environment

This issue is targeted at potential applications for Parsl's participation in Outreachy. The tasks in this issue can be performed by any number of applicants. Talk to @benclifford on Slack if you are interested.

The point of the tasks in this issue are to walk you through getting a development environment set up, making a trivial change to the code, seeing how our tooling tests things, deliberately introducing a problem, and then fixing that problem. By the end you should have a trivial 1-byte change to submit to parsl as a PR, to try out the PR process.

There are lots of steps because I have tried to make each step small.

If you get stuck or don't understand any of these steps - you can ask @benclifford on Slack or write publicly in the #parsl-hackers channel on  Slack.

1. Get a Python environment - parsl is supported on Python versions 3.8 - 3.12.
2. Sign up for a GitHub account
3. Go to the parsl repository at https://github.com/Parsl/parsl and near the top right, click "Forks / create a new fork". This will make a fork of parsl in your own github account.
5. Install parsl - follow the instructions in the "For developers" section of the Parsl README at https://github.com/Parsl/parsl - but when asked to run the command `git clone https://github.com/parsl/parsl`, instead use the address of your fork created in the previous step.
6. You can run some local tests by running `make local_thread_test`, and some more advanced tests with `make htex_local_alternate_test`
7. You can test our code linting by running `make flake8' which checks all of our Python code to see if it matches the Python style rules in [PEP-8](https://peps.python.org/pep-0008/)
8. We have configuration to let us break the style rules sometimes. This lives in the file called `.flake8`. You can open that file and make a change to move us towards proper PEP-8 compliance: find the line that says `max-line-length = 158` and make it 1 smaller. (PEP-8 says the max line length should be 80... but we can't change it all at once)
9. Run `make flake8` again - you should see that you can an error: flake8 should have found a line that is now too long.
10. Edit that line - you'll have to figure out how to make it shorter, perhaps by splitting it into two valid lines of Python
11. Run `make flake8` again and you should see that flake8 passes again
12. Now you can make a git commit: type `git commit -a -m "Reduce flake8 max-line-length`. Don't worry if you don't understand what git is doing: you can learn more about that during the internship.
13. Now push that commit up to github: type `git push`
14. Visit your repository on github in your web browser - you should see your commit there now.
15. Click "Pull request" to make a pull request to the main Parsl repository.
16. Fill out the template - you don't need to write too much
17. Sit back and wait for someone to review your pull request - that might happen within minutes, or it might take many hours, depending on who is working/awake/etc. Don't panic if you haven't heard after a few hours!
18. You might get some feedback and need to make some changes - the reviewer will help you with that
19. or... they might merge your change, in which case Congratulations! you've made a tiny contribution to Parsl, and our code base is now a tiny bit cleaner!

while setting up this project I encountered some issues. I followed the instructions here and also on the readme, below are some of the issues and how I resolved them with the help of @benclifford 

1. when I ran the `make deps`  instruction from the readme file for Build and Test

**Issue:**  `ERROR: failed building wheel for mpi4py`

**Likely reason for the issues:** mpich is not installed for Linux users

**Solution:** `sudo apt install mpich` for Ubuntu users

2. when I ran the command on #3105 No. 5

**Issues:** 
```shell
pytest parsl/tests/ -k "not cleannet" --config parsl/tests/configs/local_threads.py --random-order --durations 10
/usr/bin/bash: line 1: pytest: command not found
make: *** [Makefile:52: local_thread_test] Error 127
``` 
**Likely reason for the issues:** The `make deps` didn't execute successfully, possibly because of my internet connection

**Solution:** I tried to rerun the `make deps` command again, this time it executed successfully.

then I was able to run the `make local_thread_test`  in #3105 No. 5 again and it worked

3. when I ran the build and test command it generated some random files and folders which are unnecessary.

**Issues:** These randomly generated files and folders were added to my commit when I tried to push because I ran the 
`git add . `  command

**Solution:** I ran the `git commit -a` command instead of using the `git add .` command and it will only commit files that were already there, not any new stuff

4. From the readme on the git repo the part that says `python3 setup.py install` in step 3 didn't work for me 

**Issue:** 
```shell
/home/work/Open-Source/parsl-fork/.venv/lib/python3.10/site-packages/setuptools/dist.py:493: UserWarning: Normalizing '1.3.0-dev' to '1.3.0.dev0'
  warnings.warn(tmpl.format(**locals()))
running install
/home/work/Open-Source/parsl-fork/.venv/lib/python3.10/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
  warnings.warn(
/home/work/Open-Source/parsl-fork/.venv/lib/python3.10/site-packages/setuptools/command/easy_install.py:158: EasyInstallDeprecationWarning: easy_install command is deprecated. Use build and pip and other standards-based tools.
  warnings.warn(
running bdist_egg
running egg_info
error: [Errno 13] Permission denied
``` 

**Likely reason for the issues:**  I ran this command with sudo `make virtualenv` 

**Solution:**  I deleted the .venv folder by running  `sudo rm -r .venv`, also deleted the parsl.egg-info folder `sudo rm -r parsl.egg-info` and rerun the  `python3 setup.py install`  command


### NOTE:

if you are running a Windows OS on your PC, you might need to install WSL, this gives you an Ubuntu experience on your window machine

Read about how to install [WSL](https://learn.microsoft.com/en-us/windows/wsl/install) on your PC https://learn.microsoft.com/en-us/windows/wsl/install

> while setting up this project I encountered some issues. I followed the instructions here and also on the readme, below are some of the issues and how I resolved them with the help of @benclifford
> 
> 1. when I ran the `make deps`  instruction from the readme file for Build and Test
> 
> **Issue:** `ERROR: failed building wheel for mpi4py`
> 
> **Likely reason for the issues:** mpich is not installed for Linux users
> 
> **Solution:** `sudo apt install mpich` for Ubuntu users
> 
> 2. when I ran the command on [Outreachy: getting familiar with the development environment #3105](https://github.com/Parsl/parsl/issues/3105) No. 5
> 
> **Issues:**
> 
> ```shell
> pytest parsl/tests/ -k "not cleannet" --config parsl/tests/configs/local_threads.py --random-order --durations 10
> /usr/bin/bash: line 1: pytest: command not found
> make: *** [Makefile:52: local_thread_test] Error 127
> ```
> 
> **Likely reason for the issues:** The `make deps` didn't execute successfully, possibly because of my internet connection
> 
> **Solution:** I tried to rerun the `make deps` command again, this time it executed successfully.
> 
> then I was able to run the `make local_thread_test` in #3105 No. 5 again and it worked
> 
> 3. when I ran the build and test command it generated some random files and folders which are unnecessary.
> 
> **Issues:** These randomly generated files and folders were added to my commit when I tried to push because I ran the `git add . ` command
> 
> **Solution:** I ran the `git commit -a` command instead of using the `git add .` command and it will only commit files that were already there, not any new stuff
> 
> 4. From the readme on the git repo the part that says `python3 setup.py install` in step 3 didn't work for me
> 
> **Issue:**
> 
> ```shell
> /home/work/Open-Source/parsl-fork/.venv/lib/python3.10/site-packages/setuptools/dist.py:493: UserWarning: Normalizing '1.3.0-dev' to '1.3.0.dev0'
>   warnings.warn(tmpl.format(**locals()))
> running install
> /home/work/Open-Source/parsl-fork/.venv/lib/python3.10/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
>   warnings.warn(
> /home/work/Open-Source/parsl-fork/.venv/lib/python3.10/site-packages/setuptools/command/easy_install.py:158: EasyInstallDeprecationWarning: easy_install command is deprecated. Use build and pip and other standards-based tools.
>   warnings.warn(
> running bdist_egg
> running egg_info
> error: [Errno 13] Permission denied
> ```
> 
> **Likely reason for the issues:** I ran this command with sudo `make virtualenv`
> 
> **Solution:** I deleted the .venv folder by running `sudo rm -r .venv`, also deleted the parsl.egg-info folder `sudo rm -r parsl.egg-info` and rerun the `python3 setup.py install` command
> 
> ### NOTE:
> if you are running a Windows OS on your PC, you might need to install WSL, this gives you an Ubuntu experience on your window machine
> 
> Read about how to install [WSL](https://learn.microsoft.com/en-us/windows/wsl/install) on your PC https://learn.microsoft.com/en-us/windows/wsl/install

I hope this will be helpful to someone

> while setting up this project I encountered some issues. I followed the instructions here and also on the readme, below are some of the issues and how I resolved them with the help of @benclifford
> 
> 1. when I ran the `make deps`  instruction from the readme file for Build and Test
> 
> **Issue:** `ERROR: failed building wheel for mpi4py`
> 
> **Likely reason for the issues:** mpich is not installed for Linux users
> 
> **Solution:** `sudo apt install mpich` for Ubuntu users
> 
> 2. when I ran the command on [Outreachy: getting familiar with the development environment #3105](https://github.com/Parsl/parsl/issues/3105) No. 5
> 
> **Issues:**
> 
> ```shell
> pytest parsl/tests/ -k "not cleannet" --config parsl/tests/configs/local_threads.py --random-order --durations 10
> /usr/bin/bash: line 1: pytest: command not found
> make: *** [Makefile:52: local_thread_test] Error 127
> ```
> 
> **Likely reason for the issues:** The `make deps` didn't execute successfully, possibly because of my internet connection
> 
> **Solution:** I tried to rerun the `make deps` command again, this time it executed successfully.
> 
> then I was able to run the `make local_thread_test` in #3105 No. 5 again and it worked
> 
> 3. when I ran the build and test command it generated some random files and folders which are unnecessary.
> 
> **Issues:** These randomly generated files and folders were added to my commit when I tried to push because I ran the `git add . ` command
> 
> **Solution:** I ran the `git commit -a` command instead of using the `git add .` command and it will only commit files that were already there, not any new stuff
> 
> 4. From the readme on the git repo the part that says `python3 setup.py install` in step 3 didn't work for me
> 
> **Issue:**
> 
> ```shell
> /home/work/Open-Source/parsl-fork/.venv/lib/python3.10/site-packages/setuptools/dist.py:493: UserWarning: Normalizing '1.3.0-dev' to '1.3.0.dev0'
>   warnings.warn(tmpl.format(**locals()))
> running install
> /home/work/Open-Source/parsl-fork/.venv/lib/python3.10/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
>   warnings.warn(
> /home/work/Open-Source/parsl-fork/.venv/lib/python3.10/site-packages/setuptools/command/easy_install.py:158: EasyInstallDeprecationWarning: easy_install command is deprecated. Use build and pip and other standards-based tools.
>   warnings.warn(
> running bdist_egg
> running egg_info
> error: [Errno 13] Permission denied
> ```
> 
> **Likely reason for the issues:** I ran this command with sudo `make virtualenv`
> 
> **Solution:** I deleted the .venv folder by running `sudo rm -r .venv`, also deleted the parsl.egg-info folder `sudo rm -r parsl.egg-info` and rerun the `python3 setup.py install` command
> 
> ### NOTE:
> if you are running a Windows OS on your PC, you might need to install WSL, this gives you an Ubuntu experience on your window machine
> 
> Read about how to install [WSL](https://learn.microsoft.com/en-us/windows/wsl/install) on your PC https://learn.microsoft.com/en-us/windows/wsl/install

I hope this will be helpful to someone

--- a/.flake8
+++ b/.flake8
@@ -8,7 +8,7 @@
 # W504: line break after binary operator
 #       (Raised by flake8 even when it is followed)
 ignore = E126, E402, E129, W504
-max-line-length = 146
+max-line-length = 145
 exclude = test_import_fail.py,
   parsl/executors/workqueue/parsl_coprocess.py
 # E741 disallows ambiguous single letter names which look like numbers

--- a/parsl/executors/high_throughput/process_worker_pool.py
+++ b/parsl/executors/high_throughput/process_worker_pool.py
@@ -361,7 +361,9 @@ class Manager:
                     kill_event.set()
                 else:
                     task_recv_counter += len(tasks)
-                    logger.debug("Got executor tasks: {}, cumulative count of tasks: {}".format([t['task_id'] for t in tasks], task_recv_counter))
+                    logger.debug("Got executor tasks: {}, cumulative count of tasks: {}".format(
+                        [t['task_id'] for t in tasks], task_recv_counter
+                    ))
 
                     for task in tasks:
                         self.task_scheduler.put_task(task)

--- a/parsl/jobs/strategy.py
+++ b/parsl/jobs/strategy.py
@@ -243,7 +243,8 @@ class Strategy:
                 else:
                     # We want to make sure that max_idletime is reached
                     # before killing off resources
-                    logger.debug(f"Strategy case 1b: Executor has no active tasks, and more ({active_blocks}) than minimum blocks ({min_blocks})")
+                    logger.debug(f"Strategy case 1b: Executor has no active tasks, and more ({active_blocks})"
+                                 f" than minimum blocks ({min_blocks})")
 
                     if not self.executors[executor.label]['idle_since']:
                         logger.debug(f"Starting idle timer for executor. If idle time exceeds {self.max_idletime}s, blocks will be scaled in")
@@ -259,7 +260,8 @@ class Strategy:
 
                     else:
                         logger.debug(
-                                f"Idle time {idle_duration}s is less than max_idletime {self.max_idletime}s for executor {label}; not scaling in")
+                                f"Idle time {idle_duration}s is less than max_idletime {self.max_idletime}s"
+                                f" for executor {label}; not scaling in")
 
             # Case 2
             # More tasks than the available slots.

